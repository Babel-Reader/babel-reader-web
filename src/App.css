.react-pdf__Page__canvas {
    margin:auto;
    position: absolute;
    right: 0;
    left: 0;
}


.react-pdf__Page__textContent {
    height: 100% !important;
    position: absolute !important;
    transform: none !important;
    left: 0 !important;
    right: 0 !important;
    margin: auto !important;
}

