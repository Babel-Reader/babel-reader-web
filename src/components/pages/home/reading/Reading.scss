.reading-container {
  display: flex;
  flex-grow: 1;
  background-color: gray;
  overflow: auto;

  .nav-btn-back {
    left: 0;
    bottom: 0;
    height: 100%;
    font-size: 8vw;
    cursor: pointer;
    text-shadow: 0.4vw 0.4vw rgba(64, 62, 62, 0.25);
    position: fixed;
    z-index: 1;
  }

  .nav-btn-forward {
    right: 0;
    bottom: 0;
    height: 100%;
    font-size: 8vw;
    cursor: pointer;
    text-shadow: 0.4vw 0.4vw rgba(64, 62, 62, 0.25);
    position: fixed;
    z-index: 1;
  }

  .reading-document {
    width: 100%;

    .react-pdf__Page {
      :not([class^='translation-popup']):not(.selected) {
        :hover {
          background-color: #3399ff14;
        }

        :active {
          background-color: #3399ff64;
        }
      }

      height: 0;

      .react-pdf__Page__canvas {
        outline-color: black;
        outline-style: dashed;
      }

      .selected {
        background-color: #3399ff32;
        border-radius: 1px;
      }
    }
  }
}
