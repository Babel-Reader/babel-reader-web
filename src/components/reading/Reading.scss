
.reading-container {
  display: flex;
  flex-grow: 1;
  background-color: gray;
  overflow: auto;

  div[class^='nav-btn'] {
    text-align: center;
    display: flex;
    align-items: center;
    justify-content: center;

    width: 10%;
    position: absolute;
    z-index: 1;
    height: 100%;
    background-color: rgba(240, 240, 240, 0.5);
  }
  .nav-btn-back{
    left:0;
  }
  .nav-btn-forward{
    right: 0;
  }

  .reading-document {
    width: 100%;
    .react-pdf__Page {
      height: 0;

      .react-pdf__Page__canvas {
        outline-color: black;
        outline-style: dashed;
      }

      .selected {
        background-color: #3399ff32;
        border-radius: 1px;
      }
    }
  }
}
