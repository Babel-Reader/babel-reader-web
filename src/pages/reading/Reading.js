import {withRouter} from 'react-router-dom';
import {Document, Page} from "react-pdf";
import React from "react";
import Translate from "../../translate-api/Translate";

const Reading = ({file = 'sample.pdf', inLang = '', outLang = 'en'}) => {

  const onClick = ({target}) => {
    if (target) {
      const text = target.textContent;
      if (text) {
        Translate(text, {inLang, outLang}, (res) => {
          alert(res);
        });
      }
    }
  }

  return (
    <>
      <Document
        file={file}>
        <Page pageNumber={1} onClick={onClick}/>
      </Document>
    </>
  );
}

export default withRouter(Reading);
