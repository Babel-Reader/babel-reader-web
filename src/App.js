import React, { useEffect, useState } from 'react';
import './App.css';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import Home from './components/pages/home/Home';
import { pdfjs } from 'react-pdf';
import Library from './components/pages/library';
import { bookTitle } from 'utils';
import { auth } from 'services/firebase/firebase';

pdfjs.GlobalWorkerOptions.workerSrc = `//cdnjs.cloudflare.com/ajax/libs/pdf.js/${pdfjs.version}/pdf.worker.js`;

export const LANG_AUTO = {
  key: '??',
  name: 'auto detect',
  nativeName: 'auto detect',
};

let languageList = [
  LANG_AUTO,
  {
    key: 'en',
    name: 'English',
    nativeName: 'English',
  },
  {
    key: 'fr',
    name: 'French',
    nativeName: 'Français',
  },
  {
    key: 'es',
    name: 'Spanish',
    nativeName: 'Español',
  },
];

const fetchLanguageList = () => {
  return fetch(process.env.REACT_APP_TRANSLATE_LANGUAGE_LIST_URL)
    .then((res) => res.json())
    .then((res) => {
      if (res && res.translation) {
        const langValueForKey = (key) => {
          return {
            key,
            name: res.translation[key].name,
            nativeName: res.translation[key].nativeName,
          };
        };

        languageList = [
          LANG_AUTO,
          langValueForKey('en'),
          langValueForKey('fr'),
          langValueForKey('es'),
        ];
        Object.keys(res.translation).forEach((key) => {
          if (key !== 'en' && key !== 'fr' && key !== 'es') {
            languageList.push(langValueForKey(key));
          }
        });
      }
    });
};

export const BookContext = React.createContext({});

function App() {
  const [file, setFile] = useState();
  const [user, setUser] = useState(auth.currentUser);
  const [languages, setLanguages] = useState({
    in: LANG_AUTO,
    out: languageList[2],
  });

  useEffect(() => {
    const title = bookTitle(file);
    const pageTitle = 'Babel Reader';
    document.title = title ? `${pageTitle}: ${title}` : pageTitle;
  }, [file]);

  useEffect(() => {
    fetchLanguageList().then(() => {
      setLanguages({
        in: LANG_AUTO,
        out: languageList[2],
      });
    });
  }, []);

  return (
    <BookContext.Provider
      value={{
        file,
        setFile,
        languages,
        setLanguages,
        languageList,
        user,
        setUser,
      }}
    >
      <Router>
        <Switch>
          <Route path="/reading/samples/:bookName">
            <Home isSample />
          </Route>
          <Route path="/reading/:bookName">
            <Home />
          </Route>
          <Route>
            <Library />
          </Route>
        </Switch>
      </Router>
    </BookContext.Provider>
  );
}

export default App;
