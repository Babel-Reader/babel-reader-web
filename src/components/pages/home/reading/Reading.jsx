import { Document, Page } from 'react-pdf';
import React, { useContext, useEffect, useRef, useState } from 'react';

import './Reading.scss';
import TranslationPopup from './translation-popup/TranslationPopup';
import Button from '@material-ui/core/Button';
import translate from 'services/TranslateApi';
import { BookContext } from 'App';

const Reading = () => {
  const {
    languages,
    file,
    pageNb,
    setPageNb,
    pageCount,
    setPageCount,
    pageScale,
    setPageScale,
  } = useContext(BookContext);
  const [selected, setSelected] = useState();
  const [translation, setTranslation] = useState('');
  const ref = useRef();

  const { in: inLang, out: outLang } = languages;

  const select = (target) => {
    setTranslation('');
    if (selected) {
      selected.classList.remove('selected')
    }

    setSelected(target);
  };

  const onClick = ({ target }) => {
    let selection = null;

    if (target && target.parentElement.classList.contains('react-pdf__Page__textContent')) {
     target.classList.add("selected")
      const text = target.textContent;
      if (text) {
        selection = target;
        translate(text, { inLang, outLang }, (res) => {
          setTranslation(res);
        });
      }
    }

    select(selection);
  };

  useEffect(() => {
    setPageScale(1);
  }, [pageNb, setPageScale]);

  return (
    <div className="reading-container" ref={ref}>
      <Button
        className="nav-btn-back"
        disabled={pageNb<=1}
        onClick={() => {
          setSelected(null);
          setPageNb(pageNb - 1);
        }}
      >
        {'<'}
      </Button>
      <Document
        className="reading-document"
        onLoadSuccess={({ numPages }) => {
          setPageCount(numPages);
        }}
        file={file}
      >
        <Page
          className={`reading-page`}
          pageNumber={pageNb}
          onClick={onClick}
          scale={pageScale}
          onLoadSuccess={(page) => {
            const widthRatio = ref.current.offsetWidth / page.width || 1;
            const heightRatio = ref.current.offsetHeight / page.height || 1;

            setPageScale(Math.min(widthRatio, heightRatio));
          }}
        >
          {selected && translation && (
          <TranslationPopup selected={selected} translation={translation}>
            {translation}
          </TranslationPopup>
          )}
        </Page>
      </Document>
      <Button
        className="nav-btn-forward"
        disabled={pageNb >= pageCount}
        onClick={() => {
          setSelected(null);
          setPageNb(pageNb + 1);
        }}
      >
        {'>'}
      </Button>
    </div>
  );
};

export default Reading;
