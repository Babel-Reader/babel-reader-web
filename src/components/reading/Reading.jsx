import {withRouter} from 'react-router-dom';
import {Document, Page} from "react-pdf";
import React, {useContext, useState} from "react";
import Translate from "../../translate-api/Translate";

import './Reading.scss'
import {BookContext} from "../home/Home";
import TranslationPopup from "./translation-popup/TranslationPopup";


const Reading = () => {
  const {languages, file, pageNb, setPageNb, pageCount, setPageCount} = useContext(BookContext);
  const [selected, setSelected] = useState();
  const [translation, setTranslation] = useState('')

  const {in: inLang, out: outLang} = languages;


  const select = (target) => {
    setTranslation('')
    if (selected) {
      selected.className = selected.className.split(' selected')[0]
    }

    if (target) {
      target.className += " selected";
    }
    setSelected(target);
  }

  const onClick = ({target}) => {
    let selection = null;
    if (target) {
      const text = target.textContent;
      if (text) {
        selection = target;
        Translate(text, {inLang, outLang}, (res) => {
          setTranslation(res);
        });
      }
    }
    select(selection)
  }

  return (
    <div className='reading-container'>
      <div className='nav-btn-back'
           onClick={() => {
             setSelected(null);
             if (pageNb > 1) {
               setPageNb(pageNb - 1)
             }
           }}
      >
        {'<'}
      </div>
      <Document
        className='reading-document'
        onLoadSuccess={({numPages}) => {
          setPageCount(numPages)
        }}
        file={file}>
        <Page pageNumber={pageNb} onClick={onClick}>
          {/*todo get page width?*/}
          {selected && (
            <TranslationPopup selected={selected} translation={translation}>
              {translation}
            </TranslationPopup>
          )}
        </Page>
      </Document>
      <div className='nav-btn-forward'
           onClick={() => {
             setSelected(null);
             if (!pageCount || pageNb < pageCount) {
               setPageNb(pageNb + 1);
             }
           }}
      >
        {'>'}
      </div>
    </div>
  );
}

export default withRouter(Reading);
