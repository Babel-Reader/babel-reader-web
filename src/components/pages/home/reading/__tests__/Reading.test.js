import React from 'react';
import { create } from 'react-test-renderer';
import Reading from '../Reading';
import { BookContext } from 'App'

it('can match snapshot', () => {
  expect(
    create(
      <BookContext.Provider
        value={{
          languages: {
            in: {},
            out: {},
          },
        }}
      >
        <Reading />
      </BookContext.Provider>
    )
  ).toMatchSnapshot();
});
