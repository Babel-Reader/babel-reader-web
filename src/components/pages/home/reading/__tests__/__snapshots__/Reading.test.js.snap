// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can match snapshot 1`] = `
<div
  className="reading-container"
>
  <div
    className="nav-btn-back"
    onClick={[Function]}
  >
    &lt;
  </div>
  <div
    className="react-pdf__Document reading-document"
  >
    <div
      className="react-pdf__message react-pdf__message--no-data"
    >
      No PDF file specified.
    </div>
  </div>
  <div
    className="nav-btn-forward"
    onClick={[Function]}
  >
    &gt;
  </div>
</div>
`;
